dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/DataFileServiceBean.java

```diff
@@ -1590,23 +1568,31 @@ private File saveInputStreamInTempFile(InputStream inputStream, Long fileSizeLim
             throws IOException, FileExceedsMaxSizeException {
         Path tempFile = Files.createTempFile(Paths.get(getFilesTempDirectory()), "tmp", "upload");
 
-        if (inputStream != null && tempFile != null) {
-            Files.copy(inputStream, tempFile, StandardCopyOption.REPLACE_EXISTING);
+        FileOutputStream outStream = new FileOutputStream(tempFile.toFile());
+        logger.info("Will attempt to save the file as: " + tempFile.toString());
 
-            // size check:
-            // (note that "no size limit set" = "unlimited")
-            Long fileSize = tempFile.toFile().length();
-            if (fileSizeLimit != null && fileSize > fileSizeLimit) {
+        byte[] buffer = new byte[8 * 1024];
+        int bytesRead;
+        long totalBytesRead = 0;
+        while ((bytesRead = inputStream.read(buffer)) != -1) {
+            totalBytesRead += bytesRead;
+            if (fileSizeLimit != null && totalBytesRead > fileSizeLimit) {
                 try {
                     tempFile.toFile().delete();
                 } catch (Exception ex) {
                 }
-                throw new FileExceedsMaxSizeException(MessageFormat.format(BundleUtil.getStringFromBundle("file.addreplace.error.file_exceeds_limit"), bytesToHumanReadable(fileSize), bytesToHumanReadable(fileSizeLimit)));
-            }
+                inputStream.close();
+                outStream.close();
 
-            return tempFile.toFile();
+                throw new FileExceedsMaxSizeException(MessageFormat.format(BundleUtil.getStringFromBundle("file.addreplace.error.file_exceeds_limit"),
+                                                                           bytesToHumanReadable(fileSizeLimit)));
+            }
+            outStream.write(buffer, 0, bytesRead);
         }
-        throw new IOException("Failed to save uploaded file.");
+        inputStream.close();
+        outStream.close();
```

Comment:
Could we modify the flow here in order to have `outStream` declared inside `try-with-resources` clause? If not then we should at least place all `close()`s inside `try-finally` blocks.